ipv4-heatmap.1
.\" 
.\"   IPv4 Heatmap
.\"   (C) 2011 The Measurement Factory, Inc
.\"   Licensed under the GPL, version 2
.\"   http://maps.measurement-factory.com/
.\"
.Dd 02 May, 2011
.Dt ipv4-heatmap 1
.Os
.Sh NAME
.Nm ipv4-heatmap
.Nd Create a map of IPv4 address data
.Sh SYNOPSIS
.Nm
.Op Fl dhprmT
.Op Fl A Ar float
.Op Fl B Ar float
.Op Fl a Ar file
.Op Fl f Ar font
.Op Fl g Ar seconds
.Op Fl k Ar file
.Op Fl o Ar file
.Op Fl s Ar file
.Op Fl t Ar string
.Op Fl u Ar string
.Op Fl y Ar prefix
.Op Fl z Ar bits
< iplist
.Sh DESCRIPTION
.Nm
is a program that generates a map of IPv4 address data using a
space-filling Hilbert Curve.  Inspiration for
.Nm
comes from the xkcd comic (http://www.xkcd.org/195/).
.Pp
The output of
.Nm
is a 4096x4096 PNG image.
Each pixel in the image represents a single /24 network and is assigned
one of 256 colors.  Typically, the pixel color
represents the number of hosts within the /24 belonging to a
dataset or having some property, such as being pingable or
being the source of some traffic.
.Pp
Pixel colors range from blue (1 host) to red (256 hosts),
while black represents no data (0 hosts).  Of course, the colors
and pixel values may also be used to represent other properties of
IPv4 addresses.
.Pp
The map may be annotated by placing transparent
text labels over specific regions of address space.
.Pp
.Nm
also supports shading of areas specified by CIDR netblocks.  This is
useful to show reserved and other special address space, for example.
.Pp
.Nm
can also add an optional legend to the map.
.Pp
.Nm
uses the GD library (libgd) to create the map image.
.Sh COMMAND LINE OPTIONS
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A Ar logmin
Input data will be scaled logarithmically such that
input values less than or equal to
.Pa logmin
will be set to 1.
.It Fl B Ar logmax
Input data will be scaled logarithmically such that
input values greater than or equal to
.Pa logmax
will be set to 255.
.It Fl C
Values accumulate when using the Exact input mode.
.It Fl a Ar annotations
The
.Pa annotations
file contains a list of annotations for the map.  See ANNOTATIONS below
for the format of this file.
.It Fl c Ar color
The color of the annotations (those that appear inside the map).  Specified
as 0xRRGGBB.
.It Fl d
increase debugging levels.
.It Fl f Ar font
Specifies the font to use for the legend and annotations.  If
libgd was compiled with fontconfig support, then this can be a
fontconfig string such as "Times-12:bold".  Otherwise, you can
also specify the pathname to a True Type Font (.ttf) file.
.It Fl g Ar seconds
This option enables animated GIF output mode.  A new frame is created for
each
.Pa seconds
interval of the input data file.  See the ANIMATED GIFS below for
additional details.
.It Fl h
Attach a horizontal legend to the bottom of the map.  Note that
the legend is drawn only if the
.Fl t
option is given.
.It Fl k Ar keyfile
Use
.Pa keyfile
to create the legend scale, rather than the built-in blue-to-red scale.
.It Fl m
Use Morton (aka "Z") Curve ordering instead of Hilbert.
.It Fl o Ar outfile
Output file name.  If none is given, the image is saved as map.png by
default.
.It Fl p
Include a section in the legend that shows the size of CIDR prefixes.
Boxes and labels will be drawn to show the size of /8, /12, /16, /20, and /24
prefixes.
.It Fl r
Reverse the background and foreground colors.
.It Fl s Ar shades
The
.Ar shades
file can be used to shade certain areas of the map with specific colors
and transparency levels.  See SHADING below for the format of this file.
.It Fl T
Transpose the curve by reflecting it about its main diagonal.  This
preserves the locality properties of the curve, but moves the last
address of the rendered space from the upper right corner to the lower
left corner.
.It Fl t Ar title
Instructs
.Nm
to draw a legend for the map and place the
.Ar title
string in the top (or right) section.
You may use "\\n" to create multi-line titles.  By default the legend
will be drawn vertically and attached to the right side of the map.
Use the
.Fl h
option to create a horizontal legend instead.
.It Fl u Ar string
Instructs
.Nm
to draw a scale in the legend showing the range of colors
and their values.
.Ar string
will be placed above the scale.  Currently,
.Nm
always assumes the data represents some kind of utilization 
and prints percentages from 0 to 100% next to the scale.
.It Fl y Ar cidr
Specifies the CIDR netblock that should be rendered.  The default
is to render the entire IPv4 space (0.0.0.0/0).  The "slash" value
must be even so that the output image is square.
The
.Fl y
and
.Fl z
options together determine the size of the output image.
.It Fl z Ar bits
Specifies the number of address space bits assigned to each pixel
in the output image.  By default each pixel represents a /24 network,
which corresponds to 8 host bits (i.e., 256 hosts).  Specify 0 here
for one pixel per host address.
.El
.Sh INPUT MODES
.Nm
accepts three input modes:
.Bl -enum
.It
Increment mode.
.Pp
In this mode, the input consists of IPv4 addresses only, one
address per line.  Each address in the list increments the pixel
value for the corresponding /24.
In this mode, repeated addresses may lead to false results because
.Nm
does not check for uniqueness of the input values.  It silently
limits pixels to the maximum value.  The user should run the data
through
.Xr sort 1
and
.Xr uniq 1
beforehand, if necessary.
.It
Exact mode.
.Pp
In this mode, the input consists of two whitespace-separated fields:
an IPv4 address and a color index.  The color index is an integer
in the range 0--255.  Be careful with this mode because later
addresses may overwrite earlier ones in the same /24.
.It
Logarithmic mode.
.Pp
Very similar to Exact mode, except that the second column is
logarithmically scaled to calculate the color index.  In order to
use this mode the input must have two fields, and the
.Fl A Ar logmin
and
.Fl B Ar logmax
command line options must be given.  Color index
.Vt k
is calculated from input value
.Vt i
according to this formula:
.Bd -literal -offset indent
            ln ( i / logmin)
k = 255 * -------------------- 
          ln (logmax / logmin)
.Ed
.El
.Sh ANNOTATIONS
The annotations file consists of two or three TAB-separated fields.  The first field
is a CIDR prefix, and the second is the annotation string.  The annotation string
is rendered within
the boundary of that prefix, centered both vertically and horizontally.  
The third field, if present, is also rendered just below the 
annotation text in a fixed-size 12-point font.  If the third field is
equal to the string "prefix" then the CIDR prefix (from field 1) is rendered instead.
.Pp
The annotations file might look like this, for example:
.Bd -literal -offset indent
15.0.0.0/8      HP
16.0.0.0/8      DEC
17.0.0.0/8      Apple
.Ed
.Pp
The
.Nm
source code distribution should include a file named iana-labels.txt,
which is based on the list of IPv4 assignments made by IANA.
.Pp
The font can be selected with the
.Fl f
command line option.
At this time, however, the text color and transparency are hard-coded in the
.Nm
program.
.Sh SHADING
Areas of the map can be "shaded" by using the
.Fl s Ar shades
option.  This was originally developed simply to highlight address space that
is designated reserved or unallocated, but is also a useful way to
render complex map data.  The
.Ar shades
file consists of three TAB-separated fields: prefix, color, and alpha value.
.Pp
The CIDR prefix specifies the area to be shaded.  The color should
be specified has a hexadecimal value beginning with "0x".  The alpha value
controls the transparency of the shaded area and is passed directly to
the GD library functions.  An alpha value of 0 means full transparency, while
a value of 127 means no transparency (full opacity).  Here is an example that
shows RFC 1918 address space in a light purple color:
.Bd -literal -offset indent
10.0.0.0/8      0x7F7FFF        64
172.16.0.0/12   0x7F7FFF        64
192.168.0.0/16  0x7F7FFF        64
.Ed
.Sh ANIMATED GIFS
When the
.Fl g
option is given,
.Nm
outputs an animated GIF image file.  This feature requires the
.Xr gifsicle 1
program to be installed.
.Pp
This feature also requires timestamps in the input data.  Thus, use of the
.Fl g 
option changes the input format.  Each line of the input must begin with
a timestamp given in Unix epoch time.  For example:
.Bd -literal -offset indent
1234567890.123  192.168.1.1
1234567890.234  192.168.1.2
1234567891.456  192.168.1.3
.Ed
.Pp
Note that decimal time values are accepted, although the fractional seconds are
ignored.
.Nm
Assumes that the input timestamps are already sorted.  A new output frame
is generated every
.Ar seconds
seconds of the input file.
.Pp
Note that, currently, the data accumulates between frames.  That is, any
pixels that are colored at the end of one frame will also be colored at the
start of the next frame.
.Sh HILBERT CURVE
.Nm
uses a 12th-order Hilbert Curve to represnet the entire IPv4 address
space.  Locating a particular IP address along the curve can
be confusing at first.  
Here is what a 2nd-order Hilbert curve looks like:
.Bd -literal -offset indent
    0---1   14--15
        |   |
    3---2   13--12
    |            |
    4   7---8   11
    |   |   |    |
    5---6   9---10
.Ed
.Pp
The best way to understand how the Hilbert Curve works is to try
drawing your own!
.Sh COPYRIGHT
.Bd -literal -offset indent
IPv4 Heatmap
(C) 2011 The Measurement Factory, Inc
Licensed under the GPL, version 2
http://maps.measurement-factory.com/
.Ed
.Sh SEE ALSO
.Xr gifsicle 1
.Sh AUTHORS
.Nm
was written by Duane Wessels of The Measurement Factory, Inc.
With contributions from: Roy Arends of Nominet UK.
.Sh BUGS
Can't draw IPv6 address maps.
.Pp
The legends don't look all that great.  You can use an image editing
program like The Gimp to rearrange the legend and add better-looking
text.


ipv4-heatmap.c
/*
 * IPv4 Heatmap
 * (C) 2007 The Measurement Factory, Inc
 * Licensed under the GPL, version 2
 * http://maps.measurement-factory.com/
 */

/*
 * ipv4-heatmap produces a "map" of IPv4 address space.
 * 
 * input is a list of IPv4 addrs and optional value for each.
 * 
 * Data is drawn using a hilbert curve, which preserves grouping see
 * http://xkcd.com/195/ and http://en.wikipedia.org/wiki/Hilbert_curve see
 * Hacker's Delight (Henry S. Warren, Jr. 2002), sec 14-2, fig 14-5
 * 
 * output is a squre PNG file
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <assert.h>
#include <math.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>


#include <gd.h>
#include <gdfonts.h>

#include "hsv2rgb.h"
#include "cidr.h"
#include "ipv4-heatmap.h"
#include "annotate.h"
#include "shade.h"
#include "legend.h"
#include "xy_from_ip.h"

#define NUM_DATA_COLORS 256
#undef RELEASE_VER

gdImagePtr image = NULL;
int colors[NUM_DATA_COLORS];
int num_colors = NUM_DATA_COLORS;
int debug = 0;
const char *whitespace = " \t\r\n";
const char *font_file_or_name = "Luxi Mono:style=Regular";
const char *legend_orient = "vert";
const char *annotations = NULL;
const char *shadings = NULL;
const char *title = NULL;
const char *legend_scale_name = NULL;
int legend_prefixes_flag = 0;
int reverse_flag = 0;		/* reverse background/font colors */
int morton_flag = 0;
int accumulate_counts = 0;	/* for when the input data contains a value */
struct {
	unsigned int secs;
	double input_time;
	time_t next_output;
} anim_gif = {0, 0.0, 0};
const char *legend_keyfile = NULL;
const char *savename = "map.png";

static void savegif(int done);
static void annotate(gdImagePtr);

/*
 * if log_A and log_B are set, then the input data will be scaled
 * logarithmically such that log_A -> 0 and log_B -> 255. log_C is calculated
 * such that log_B -> 255.
 */
double log_A = 0.0;
double log_B = 0.0;
double log_C = 0.0;

static void
initialize(void)
{
    int i;
    int w;
    int h;
    int order = set_order();
    w = 1<<order;
    h = 1<<order;
    if (title && 4096 != w) {
	warnx("Image width/height must be 4096 to render a legend.");
	fprintf(stderr,
		"\nIf you are using the -y or -z options, then your image size "
		"may be smaller\n(or larger) than 4096.  The legend-rendering "
		"code has a number of hard-coded\nparameters designed to work "
		"with a 4096x4096 output image.\n");
	exit(1);
    }
    if (NULL == title)
	(void)0;		/* no legend */
    else if (0 == strcmp(legend_orient, "horiz"))
	h += (h>>2);
    else
	w += (w>>2);
    if (debug) {
	fprintf(stderr, "image width = %d\n", w);
	fprintf(stderr, "image height = %d\n", h);
    }
    image = gdImageCreateTrueColor(w, h);
    if (image == NULL)
	err(1, "gdImageCreateTrueColor(w=%d, h=%d)", w, h);
    /* first allocated color becomes background by default */
    if (reverse_flag)
	gdImageFill(image, 0, 0, gdImageColorAllocate(image, 255, 255, 255));

    /*
     * The default color map ranges from red to blue
     */
    for (i = 0; i < NUM_DATA_COLORS; i++) {
	double hue;
	double r, g, b;
	hue = 240.0 * (255 - i) / 255;
	PIX_HSV_TO_RGB_COMMON(hue, 1.0, 1.0, r, g, b);
	colors[i] = gdImageColorAllocate(image, r, g, b);
	if (debug > 1)
	    fprintf(stderr, "colors[%d]=%d\n", i, colors[i]);
    }

    /*
     * If the input data should be logarithmically scaled, then calculate the
     * value of log_C.
     */
    if (0.0 != log_A && 0.0 == log_B)
	log_B = 10.0 * log_A;
    log_C = 255.0 / log(log_B / log_A);
}

static int
get_pixel_value(unsigned int x, unsigned int y)
{
    int color;
    int k;
    color = gdImageGetPixel(image, x, y);
    if (debug)
	fprintf(stderr, "pixel (%d,%d) has color index %d\n", x, y, color);
    for (k = 0; k < NUM_DATA_COLORS; k++) {
	if (colors[k] == color) {
	    if (debug)
		fprintf(stderr, "color %d has index %d\n", color, k);
	    break;
	}
    }
    if (k == NUM_DATA_COLORS)	/* not found */
	k = 0;
    return k;
}

void
paint(void)
{
    char buf[512];
    unsigned int line = 1;
    while (fgets(buf, 512, stdin)) {
	unsigned int i;
	unsigned int x;
	unsigned int y;
	int color = -1;
	int k;
	char *strtok_arg = buf;
	char *t;

	/*
	 * In animated gif mode the first field is a timestamp
	 */
	if (anim_gif.secs) {
	    char *e;
	    t = strtok(strtok_arg, whitespace);
	    strtok_arg = NULL;
	    if (NULL == t)
		continue;
	    anim_gif.input_time = strtod(t, &e);
	    if (e == t)
		errx(1, "bad input parsing time on line %d: %s", line, t);
	}

	/*
	 * next field is an IP address.  We also accept its integer notation
	 * equivalent.
	 */
	t = strtok(strtok_arg, whitespace);
	strtok_arg = NULL;
	if (NULL == t)
	    continue;
	if (strspn(t, "0123456789") == strlen(t))
	    i = strtoul(t, NULL, 10);
	else if (1 == inet_pton(AF_INET, t, &i))
	    i = ntohl(i);
	else
	    errx(1, "bad input parsing IP on line %d: %s", line, t);

	if (0 == xy_from_ip(i, &x, &y))
	    continue;
	if (debug)
	    fprintf(stderr, "%s => %u => (%d,%d)\n", t, i, x, y);

	/*
	 * next field is an optional value, which might also be
	 * logarithmically scaled by us.  If no value is given, then find the
	 * existing value at that point and increment by one.
	 */
	t = strtok(NULL, whitespace);
	if (NULL != t) {
	    k = atoi(t);
	    if (accumulate_counts)
		k += get_pixel_value(x, y);
	    if (0.0 != log_A) {
		/*
		 * apply logarithmic stretching
		 */
		k = (int) ((log_C * log((double) k / log_A)) + 0.5);
	    }
	} else {
	    k = get_pixel_value(x, y);
	    k++;
	}
	if (k < 0)
	    k = 0;
	if (k >= NUM_DATA_COLORS)
	    k = NUM_DATA_COLORS - 1;
	color = colors[k];

	/*
	 * Now that we're doing parsing the entire input line, we can check if
	 * an animated gif file needs to be written out.  This is done here because
         * saving the gif image can call annotation routines that also use strtok().
	 */
	if (anim_gif.secs) {
	    if ((time_t) anim_gif.input_time > anim_gif.next_output) {
		savegif(0);
		anim_gif.next_output = (time_t) anim_gif.input_time + anim_gif.secs;
	    }
	}

	gdImageSetPixel(image, x, y, color);
	line++;
    }
}

void
watermark(gdImagePtr i)
{
    int color = gdImageColorAllocateAlpha(i, 127, 127, 127, 63);
    gdImageStringUp(i,
	gdFontGetSmall(),
	gdImageSX(i) - 20, 220,
	(u_char *) "IPv4 Heatmap / Measurement Factory", color);
}

void
save(void)
{
    FILE *pngout = fopen(savename, "wb");
    if (NULL == pngout)
	err(1, "%s", savename);
    annotate(image);
    gdImagePng(image, pngout);
    fclose(pngout);
    gdImageDestroy(image);
    image = NULL;
}

void
savegif(int done)
{
	static int ngif = 0;
	static char *tdir = NULL;
	static char tmpl[] = "heatmap-tmp-XXXXXX";
	char fname[512];
	FILE *gifout = NULL;
	gdImagePtr clone;
	if (NULL == tdir) {
		tdir = mkdtemp(tmpl);
		if (NULL == tdir)
			err(1, "%s", tmpl);
	}
	snprintf(fname, 512, "%s/%07d.gif", tdir, ngif++);
	gifout = fopen(fname, "wb");
	if (NULL == gifout)	
		err(1, "%s", fname);
	clone = gdImageClone(image);
	if (NULL == clone)
		errx(1, "gdImageClone() failed");
	annotate(clone);
	gdImageGif(clone, gifout);
	fclose(gifout);
	gdImageDestroy(clone);
	/* don't destroy image! */
	if (done) {
		char cmd[512];
		snprintf(cmd, 512, "gifsicle --colors 256 %s/*.gif > %s", tdir, savename);
		fprintf(stderr, "Executing: %s\n", cmd);
		if (0 != system(cmd))
			errx(1, "gifsicle failed");
		snprintf(cmd, 512, "rm -rf %s", tdir);
		fprintf(stderr, "Executing: %s\n", cmd);
		system(cmd);
		tdir = NULL;
		gdImageDestroy(image);
		image = NULL;
	}
}

static void
annotate(gdImagePtr i)
{
    if (shadings)
	shade_file(i, shadings);
    if (annotations)
	annotate_file(i, annotations);
    if (title)
	legend(i, title, legend_orient);
    watermark(i);
}

void
usage(const char *argv0)
{
    const char *t = strrchr(argv0, '/');
    printf("IPv4 Heatmap"
#ifdef RELEASE_VER
	" (release " RELEASE_VER ")"
#endif
	"\n");
    printf("(C) 2007 The Measurement Factory, Inc\n");
    printf("Licensed under the GPL, version 2\n");
    printf("http://maps.measurement-factory.com/\n");
    printf("\n");
    printf("usage: %s [options] < iplist\n", t ? t + 1 : argv0);
    printf("\t-A float   logarithmic scaling, min value\n");
    printf("\t-B float   logarithmic scaling, max value\n");
    printf("\t-C         values accumulate in Exact input mode\n");
    printf("\t-a file    annotations file\n");
    printf("\t-c color   color of annotations (0xRRGGBB)\n");
    printf("\t-d         increase debugging\n");
    printf("\t-f font    fontconfig name or .ttf file\n");
    printf("\t-g secs    make animated gif from each secs of data\n");
    printf("\t-h         draw horizontal legend instead\n");
    printf("\t-k file    key file for legend\n");
    printf("\t-m         use morton order instead of hilbert\n");
    printf("\t-o file    output filename\n");
    printf("\t-p         show size of prefixes in legend\n");
    printf("\t-r         reverse; white background, black text\n");
    printf("\t-s file    shading file\n");
    printf("\t-T         transpose; last address in lower left, not upper right\n");
    printf("\t-t str     map title\n");
    printf("\t-u str     scale title in legend\n");
    printf("\t-y cidr    address space to render\n");
    printf("\t-z bits    address space bits per pixel\n");
    exit(1);
}

int
main(int argc, char *argv[])
{
    int ch;
    while ((ch = getopt(argc, argv, "A:B:a:Cc:df:g:hk:mo:prs:t:u:y:z:T")) != -1) {
	switch (ch) {
	case 'A':
	    log_A = atof(optarg);
	    break;
	case 'B':
	    log_B = atof(optarg);
	    break;
	case 'C':
	    accumulate_counts = 1;
	    break;
	case 'd':
	    debug++;
	    break;
	case 'a':
	    annotations = strdup(optarg);
	    break;
	case 'c':
	    annotateColor = strtol(optarg, NULL, 16);
	    break;
	case 's':
	    shadings = strdup(optarg);
	    break;
	case 'f':
	    font_file_or_name = strdup(optarg);
	    break;
	case 'g':
	    anim_gif.secs = strtol(optarg, NULL, 10);
	    break;
	case 'h':
	    legend_orient = "horiz";
	    break;
	case 'k':
	    legend_keyfile = strdup(optarg);
	    break;
	case 'o':
	    savename = strdup(optarg);
	    break;
	case 'm':
		morton_flag = 1;
		set_morton_mode();
		break;
	case 'T':
	    set_transpose_mode();
	    break;
	case 't':
	    title = strdup(optarg);
	    break;
	case 'p':
	    legend_prefixes_flag = 1;
	    break;
	case 'u':
	    legend_scale_name = strdup(optarg);
	    break;
	case 'r':
	    reverse_flag = 1;
	    break;
	case 'y':
	    set_crop(optarg);
	    break;
	case 'z':
	    set_bits_per_pixel(strtol(optarg, NULL, 10));
	    break;
	default:
	    usage(argv[0]);
	    break;
	}
    }
    argc -= optind;
    argv += optind;

    initialize();
    paint();
    if (anim_gif.secs) {
	savegif(1);
    } else {
	annotate(image);
    	save();
    }
    return 0;
}
